### PR TITLE
[Utilities] refactor promote_operation

### LINKED ISSUE #2203: Re-factor promote_operation and operate
There are a bunch of rules for various `promote_operation` and `operate`. 
https://github.com/jump-dev/MathOptInterface.jl/blob/6b2192a991bab65a087284fcba6493de09bf786d/src/Utilities/functions.jl#L1597-L1610
https://github.com/jump-dev/MathOptInterface.jl/blob/6b2192a991bab65a087284fcba6493de09bf786d/src/Utilities/functions.jl#L1697-L1715

Defining a new function is really hard because it requires whack-a-mole: https://github.com/jump-dev/MathOptInterface.jl/pull/2201

I think we need an explicit write up of what functions are implemented, along with tests, so that we can check everything is working as intended.

### PR DESCRIPTION
**Pull request by odow**
This PR consolidates the `Utilities.promote_operation` methods into a single file with tests.

<s>This is going to be a WIP until I settle on a nice design.</s>

### CODE REVIEW

**Comment by odow** [comment]
This is looking okay.

I'll tackle the `operate!` issue in a separate PR, which is also blocked by https://github.com/jump-dev/MathOptInterface.jl/pull/2207.

File: src/Utilities/promote_operation.jl [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+of the arguments `args` are `ArgsTypes`.
+
+There are five methods for which we implement `Utilities.promote_operation`:
+
+ 1. `+`
+    a. `promote_operation(::typeof(+), ::Type{T}, ::Type{F1}, ::Type{F2})`
+ 2. `-`
+    a. `promote_operation(::typeof(-), ::Type{T}, ::Type{F})`
+    b. `promote_operation(::typeof(-), ::Type{T}, ::Type{F1}, ::Type{F2})`
+    c. `promote_operation(::typeof(-), ::Type{T}, ::Type{F1}, ::Type{Vector{T}})`
```
Comment (blegat) [inline src/Utilities/promote_operation.jl:24]: Why not 
```julia
`promote_operation(::typeof(-), ::Type{T}, ::Type{F1}, ::Type{T})`
```
then ?

File: src/Utilities/promote_operation.jl [hunk 2]
```diff
@@ -0,0 +1,426 @@
...
+ 3. `*`
+    a. `promote_operation(::typeof(*), ::Type{T}, ::Type{T}, ::Type{F})`
+    b. `promote_operation(::typeof(*), ::Type{T}, ::Type{T}, ::Type{F})`
```
Comment (blegat) [inline src/Utilities/promote_operation.jl:27]:
```suggestion
    b. `promote_operation(::typeof(*), ::Type{T}, ::Type{F}, ::Type{T})`
```

File: src/Utilities/promote_operation.jl (hunk 1 above)
Comment (odow) [reply src/Utilities/promote_operation.jl]: We actually did't need 2c as a case. It was already covered by 2b.

**Comment by odow** [comment]
Merging because this is clearly a win

**Pull request merged**